fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -989,64 +1013,119 @@ void MultiVersionApi::setCallbacksOnExternalThreads() {
+	threadCount = std::max(threadCount, 1);
 
-	if(filename.empty() || !fileExists(path)) {
+	if (filename.empty() || !fileExists(path)) {
 		TraceEvent("ExternalClientNotFound").detail("LibraryPath", filename);
 		throw file_not_found();
 	}
 
-	MutexHolder holder(lock);
-	if(networkStartSetup) {
-		throw invalid_option(); // SOMEDAY: it might be good to allow clients to be added after the network is setup
-	}
-
-	if(externalClients.count(filename) == 0) {
+	if (externalClientDescriptions.count(filename) == 0) {
 		TraceEvent("AddingExternalClient").detail("LibraryPath", filename);
-		externalClients[filename] = Reference<ClientInfo>(new ClientInfo(new DLApi(path), path));
+		externalClientDescriptions.emplace(std::make_pair(filename, ClientDesc(path, true)));
 	}
 }
 
 void MultiVersionApi::addExternalLibraryDirectory(std::string path) {
 	TraceEvent("AddingExternalClientDirectory").detail("Directory", path);
 	std::vector<std::string> files = platform::listFiles(path, DYNAMIC_LIB_EXT);
-
-	MutexHolder holder(lock);
-	if(networkStartSetup) {
-		throw invalid_option(); // SOMEDAY: it might be good to allow clients to be added after the network is setup. For directories, we can monitor them for the addition of new files.
-	}
+	// we need at least one external library thread to run these libraries.
+	threadCount = std::max(threadCount, 1);
 
 	for(auto filename : files) {
 		std::string lib = abspath(joinPath(path, filename));
-		if(externalClients.count(filename) == 0) {
+		if (externalClientDescriptions.count(filename) == 0) {
 			TraceEvent("AddingExternalClient").detail("LibraryPath", filename);
-			externalClients[filename] = Reference<ClientInfo>(new ClientInfo(new DLApi(lib), lib));
-		}	
+			externalClientDescriptions.emplace(std::make_pair(filename, ClientDesc(lib, true)));
+		}
 	}
 }
+#if defined(__unixish__)
+std::vector<std::pair<std::string, bool>> MultiVersionApi::copyExternalLibraryPerThread(std::string path) {
+	ASSERT_GE(threadCount, 1);
+	// Copy library for each thread configured per version
+	std::vector<std::pair<std::string, bool>> paths;
+	// It's tempting to use the so once without copying.  However, we don't know
+	// if the thing we're about to copy is the shared object executing this code
+	// or not, so this optimization is unsafe.
+	// paths.push_back({path, false});
+	for (int ii = 0; ii < threadCount; ++ii) {
+		std::string filename = basename(path);
+
+		char tempName[PATH_MAX + 12];
+		sprintf(tempName, "/tmp/%s-XXXXXX", filename.c_str());
+		int tempFd = mkstemp(tempName);
+		int fd;
+
+		if ((fd = open(path.c_str(), O_RDONLY)) == -1) {
+			TraceEvent("ExternalClientNotFound").detail("LibraryPath", path);
+			throw file_not_found();
+		}
+
+		constexpr size_t buf_sz = 4096;
+		char buf[buf_sz];
+		while (1) {
+			ssize_t readCount = read(fd, buf, buf_sz);
+			if (readCount == 0) {
+				// eof
+				break;
+			}
+			if (readCount == -1) {
+				throw platform_error;
```

Comment:
This should be `platform_error()`. We should probably also log a note about what failed.